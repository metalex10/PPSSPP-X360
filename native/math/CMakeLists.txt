set(SRCS
  lin/matrix4x4.cpp
  lin/vec3.cpp
  lin/quat.cpp
  lin/aabb.cpp
	curves.cpp
	math_util.cpp
  )

set(SRCS ${SRCS})

add_library(lin STATIC ${SRCS})

if(UNIX)
  add_definitions(-fPIC)
endif(UNIX)
