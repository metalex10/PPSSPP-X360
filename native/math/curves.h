#pragma once

#include <base/basictypes.h>

// Easy curve computation for fades etc.


// output range: [0.0, 1.0]
float linearInOut(int t, int fadeInLength, int solidLength, int fadeOutLength);
float linearIn(int t, int fadeInLength);
float linearOut(int t, int fadeInLength);

// smooth operator [0, 1] -> [0, 1]
float ease(float val);
float ease(int t, int fadeLength);

// need a bouncy ease

// waveforms [0, 1]
float sawtooth(int t, int period);

// output range: -1.0 to 1.0
float passWithPause(int t, int fadeInLength, int pauseLength, int fadeOutLength);
