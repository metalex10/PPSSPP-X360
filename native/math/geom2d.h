#pragma once

#include <cmath>

struct Point {
	Point() {}
	Point(float x_, float y_) : x(x_), y(y_) {}

	float x;
	float y;

	float distanceTo(const Point &other) const {
		float dx = other.x - x, dy = other.y - y;
		return sqrtf(dx*dx + dy*dy);
	}

	/*
	FocusDirection directionTo(const Point &other) const {
		int angle = atan2f(other.y - y, other.x - x) / (2 * M_PI) - 0.125;

	}*/
};


// Resolved bounds on screen after layout.
struct Bounds {
	Bounds() : x(0), y(0), w(0), h(0) {}
	Bounds(float x_, float y_, float w_, float h_) : x(x_), y(y_), w(w_), h(h_) {}

	bool Contains(float px, float py) const {
		return (px >= x && py >= y && px < x + w && py < y + h);
	}

	bool Intersects(const Bounds &other) const {
		return !(x > other.x2() || x2() < other.x || y > other.y2() || y2() < other.y);
	}

	void Clip(const Bounds &clipTo) {
		if (x < clipTo.x) {
			w -= clipTo.x - x;
			x = clipTo.x;
		}
		if (y < clipTo.y) {
			h -= clipTo.y - y;
			y = clipTo.y;
		}
		if (x2() > clipTo.x2()) {
			w = clipTo.x2() - x;
		}
		if (y2() > clipTo.y2()) {
			h = clipTo.y2() - y;
		}
	}

	float x2() const { return x + w; }
	float y2() const { return y + h; }
	float centerX() const { return x + w * 0.5f; }
	float centerY() const { return y + h * 0.5f; }
	Point Center() const {
		return Point(centerX(), centerY());
	}

	float x;
	float y;
	float w;
	float h;
};


