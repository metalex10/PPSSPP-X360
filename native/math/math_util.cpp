#include "math/math_util.h"
#include <stdlib.h>

#if defined(__ARM_ARCH_7A__) && !defined(BLACKBERRY)

void EnableFZ()
{
	int x;
	asm(
		"fmrx %[result],FPSCR \r\n"
		"orr %[result],%[result],#16777216 \r\n"
		"fmxr FPSCR,%[result]"
		:[result] "=r" (x) : :
	);
	//printf("ARM FPSCR: %08x\n",x);
}
#else
void EnableFZ()
{
	// TODO
}
#endif
