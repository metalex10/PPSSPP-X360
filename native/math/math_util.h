#ifndef _MATHUTILS_H
#define _MATHUTILS_H

#include <cmath>
#include <cstring>

#include "base/basictypes.h"

inline float sqr(float f)	{return f*f;}
inline float sqr_signed(float f) {return f<0 ? -f*f : f*f;}

typedef unsigned short float16;

// This ain't a 1.5.10 float16, it's a stupid hack format where we chop 16 bits off a float.
// This choice is subject to change. Don't think I'm using this for anything at all now anyway.
// DEPRECATED
inline float16 FloatToFloat16(float x) {
	int ix;
	memcpy(&ix, &x, sizeof(float));
	return ix >> 16;
}

inline float Float16ToFloat(float16 ix) {
	float x;
	memcpy(&x, &ix, sizeof(float));
	return x;
}

inline uint32_t RoundUpToPowerOf2(uint32_t v) {
	v--;
	v |= v >> 1;
	v |= v >> 2;
	v |= v >> 4;
	v |= v >> 8;
	v |= v >> 16;
	v++;
	return v;
}

#define PI 3.141592653589793f
#ifndef M_PI
#define M_PI 3.141592653589793f
#endif

// The stuff in this file is from all over the web, esp. dspmusic.org. I think it's all public domain.
// In any case, very little of it is used anywhere at the moment.

// Calculate pseudo-random 32 bit number based on linear congruential method. 
void SetSeed(unsigned int seed);
unsigned int GenerateRandomNumber();
inline float GenerateRandomFloat01() {
	return (float)((double)GenerateRandomNumber() / 0xFFFFFFFF);
}
inline float GenerateRandomSignedFloat() {
	return (float)((double)GenerateRandomNumber() / 0x80000000) - 1.0f;
}


inline float GaussRand()
{
	float R1 = GenerateRandomFloat01();
	float R2 = GenerateRandomFloat01();

	float X = sqrtf(-2.0f * logf(R1)) * cosf(2.0f * PI * R2);
	if (X > 4.0f) X = 4.0f;
	if (X < -4.0f) X = -4.0f;
	return X;
}

// Accuracy unknown
inline double atan_fast(double x) {
	return (x / (1.0 + 0.28 * (x * x)));
}


// linear -> dB conversion
inline float lin2dB(float lin) {
	const float LOG_2_DB = 8.6858896380650365530225783783321f;	// 20 / ln( 10 )
	return logf(lin) * LOG_2_DB;
}

// dB -> linear conversion
inline float dB2lin(float dB) {
	const float DB_2_LOG = 0.11512925464970228420089957273422f;	// ln( 10 ) / 20
	return expf(dB * DB_2_LOG);
}

inline bool my_isinf(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	return f2u.u == 0x7f800000 ||
		f2u.u == 0xff800000;
}

inline bool my_isnan(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	// NaNs have non-zero mantissa
	return ((f2u.u & 0x7F800000) == 0x7F800000) && (f2u.u & 0x7FFFFF);
}

inline bool my_isnanorinf(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	// NaNs have non-zero mantissa, infs have zero mantissa. That is, we just ignore the mantissa here.
	return ((f2u.u & 0x7F800000) == 0x7F800000);
}

inline int is_even(float d) {
	float int_part;
	modff(d / 2.0f, &int_part);
	return 2.0f * int_part == d;
}

// Rounds *.5 to closest even number
inline double round_ieee_754(double d) {
	float i = floor(d);
	d -= i;
	if (d < 0.5f)
		return i;
	if (d > 0.5f)
		return i + 1.0f;
	if (is_even(i))
		return i;
	return i + 1.0f;
}

union FP32 {
	uint32_t u;
	float f;
};

struct FP16 {
	uint16_t u;
};

// magic code from ryg: http://fgiesen.wordpress.com/2012/03/28/half-to-float-done-quic/
// See also SSE2 version: https://gist.github.com/rygorous/2144712
inline FP32 half_to_float_fast5(FP16 h)
{
	static const FP32 magic = { (127 + (127 - 15)) << 23 };
	static const FP32 was_infnan = { (127 + 16) << 23 };
	FP32 o;
	o.u = (h.u & 0x7fff) << 13;     // exponent/mantissa bits
	o.f *= magic.f;                 // exponent adjust
	if (o.f >= was_infnan.f)        // make sure Inf/NaN survive
		o.u |= 255 << 23;
	o.u |= (h.u & 0x8000) << 16;    // sign bit
	return o;
}

inline float ExpandHalf(uint16_t half) {
	FP16 fp16;
	fp16.u = half;
	FP32 fp = half_to_float_fast5(fp16);
	return fp.f;
}

// More magic code: https://gist.github.com/rygorous/2156668
inline FP16 float_to_half_fast3(FP32 f)
{
	static const FP32 f32infty = { 255 << 23 };
	static const FP32 f16infty = { 31 << 23 };
	static const FP32 magic = { 15 << 23 };
	static const uint32_t sign_mask = 0x80000000u;
	static const uint32_t round_mask = ~0xfffu;
	FP16 o = { 0 };

	uint32_t sign = f.u & sign_mask;
	f.u ^= sign;

	if (f.u >= f32infty.u) // Inf or NaN (all exponent bits set)
		o.u = (f.u > f32infty.u) ? 0x7e00 : 0x7c00; // NaN->qNaN and Inf->Inf
	else // (De)normalized number or zero
	{
		f.u &= round_mask;
		f.f *= magic.f;
		f.u -= round_mask;
		if (f.u > f16infty.u) f.u = f16infty.u; // Clamp to signed infinity if overflowed

		o.u = f.u >> 13; // Take the bits!
	}

	o.u |= sign >> 16;
	return o;
}

inline uint16_t ShrinkToHalf(float full) {
	FP32 fp32;
	fp32.f = full;
	FP16 fp = float_to_half_fast3(fp32);
	return fp.u;
}

// FPU control.
void EnableFZ();

#endif
