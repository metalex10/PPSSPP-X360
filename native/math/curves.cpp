
#include <math.h>

#include "math/math_util.h"
#include "curves.h"

float linearInOut(int t, int fadeInLength, int solidLength, int fadeOutLength) {
	if (t < 0) return 0;
	if (t < fadeInLength) {
		return (float)t / fadeInLength;
	}
	t -= fadeInLength;
	if (t < solidLength) {
		return 1.0f;
	}
	t -= solidLength;
	if (t < fadeOutLength) {
		return 1.0f - (float)t / fadeOutLength;
	}
	return 0.0f;
}

float linearIn(int t, int fadeInLength) {
	if (t < 0) return 0;
	if (t < fadeInLength) {
		return (float)t / fadeInLength;
	}
	return 1.0f;
}

float linearOut(int t, int fadeOutLength) {
	return 1.0f - linearIn(t, fadeOutLength);
}

float ease(float val) {
	if (val > 1.0f) return 1.0f;
	if (val < 0.0f) return 0.0f;
	return ((-cosf(val * PI)) + 1.0f) * 0.5;
}

float ease(int t, int fadeLength)
{
	if (t < 0) return 0.0f;
	if (t >= fadeLength) return 1.0f;
	return ease((float)t / (float)fadeLength);
}

float sawtooth(int t, int period) {
	return (t % period) * (1.0f / (period - 1));
}

float passWithPause(int t, int fadeInLength, int pauseLength, int fadeOutLength)
{
	if (t < fadeInLength) {
		return -1.0f + (float)t / fadeInLength;
	}
	t -= fadeInLength;
	if (t < pauseLength) {
		return 0.0f;
	}
	t -= pauseLength;
	if (t < fadeOutLength) {
		return (float)t / fadeOutLength;
	}
	return 1.0f;
}
