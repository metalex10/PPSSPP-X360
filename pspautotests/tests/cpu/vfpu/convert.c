#include <common.h>

#include <pspkernel.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/glut.h>
//#include "../common/emits.h"

#include <pspgu.h>
#include <pspgum.h>

#include "vfpu_common.h"

__attribute__ ((aligned (16))) ScePspFVector4 v0, v1, v2;
__attribute__ ((aligned (16))) ScePspFVector4 matrix[4];

void __attribute__((noinline)) vfim(ScePspFVector4 *v0) {
	asm volatile (
		"vfim.s	 s500, 0.011111111111111112\n"
		"vfim.s	 s501, -0.011111111111111112\n"
		"vfim.s	 s502, inf\n"
		"vfim.s	 s503, nan\n"
		"sv.q    C500, %0\n"

		: "+m" (*v0)
		);
}

void __attribute__((noinline)) vf2h(ScePspFVector4 *v0, ScePspFVector4 *v1) {
	asm volatile (
		"lv.q   C100, %1\n"
		"vf2h.q C200, C100\n"
		"sv.q   C200, %0\n"

		: "+m" (*v0) : "m" (*v1)
		);
}

void __attribute__((noinline)) vh2f(ScePspFVector4 *v0, ScePspFVector4 *v1) {
	asm volatile (
		"lv.q   C100, %1\n"
		"vh2f.p C200, C100\n"
		"sv.q   C200, %0\n"

		: "+m" (*v0) : "m" (*v1)
		);
}

void __attribute__((noinline)) vf2id(ScePspFVector4 *v0, ScePspFVector4 *v1) {
	asm volatile (
		"lv.q   C100, %1\n"
		"vf2id.q C200, C100, 0\n"
		"sv.q   C200, %0\n"

		: "+m" (*v0) : "m" (*v1)
		);
}

void __attribute__((noinline)) vf2in(ScePspFVector4 *v0, ScePspFVector4 *v1) {
	asm volatile (
		"lv.q   C100, %1\n"
		"vf2in.q C200, C100, 0\n"
		"sv.q   C200, %0\n"

		: "+m" (*v0) : "m" (*v1)
		);
}

void __attribute__((noinline)) vf2iu(ScePspFVector4 *v0, ScePspFVector4 *v1) {
	asm volatile (
		"lv.q   C100, %1\n"
		"vf2iu.q C200, C100, 0\n"
		"sv.q   C200, %0\n"

		: "+m" (*v0) : "m" (*v1)
		);
}

void __attribute__((noinline)) vf2iz(ScePspFVector4 *v0, ScePspFVector4 *v1) {
	asm volatile (
		"lv.q   C100, %1\n"
		"vf2iz.q C200, C100, 0\n"
		"sv.q   C200, %0\n"

		: "+m" (*v0) : "m" (*v1)
		);
}

void __attribute__((noinline)) vi2f(ScePspFVector4 *v0, int *v1) {
	asm volatile (
		"lv.q   C100, %1\n"
		"vi2f.q C200, C100, 0\n"
		"sv.q   C200, %0\n"

		: "+m" (*v0) : "m" (*v1)
		);
}


void checkHalf() {
	static __attribute__ ((aligned (16))) ScePspFVector4 vIn1 =
	{0.9f, 1.3f, 2.7f, 1000.5f};
	static __attribute__ ((aligned (16))) ScePspFVector4 vIn2 =
	{-0.9f, -1.3f, -2.7f, -1000.5f};
	static __attribute__ ((aligned (16))) ScePspFVector4 vIn3 =
	{1.5f, 2.5f, -3.5f, -4.5f};
	static __attribute__ ((aligned (16))) ScePspFVector4 vIn4 =
	{3.5f, INFINITY, -INFINITY, NAN};

	static __attribute__ ((aligned (16))) ScePspFVector4 vOutF =
	{0.0f, 0.0f, 0.0f, 0.0f};

	struct {unsigned int x,y,z,w;} vOut;

	vf2h(&vOutF, &vIn1); memcpy(&vOut, &vOutF, 16);
	printf("vf2h: %08x,%08x,%08x,%08x\n", vOut.x, vOut.y, vOut.z, vOut.w);
	vf2h(&vOutF, &vIn2); memcpy(&vOut, &vOutF, 16);
	printf("vf2h: %08x,%08x,%08x,%08x\n", vOut.x, vOut.y, vOut.z, vOut.w);
	vf2h(&vOutF, &vIn3); memcpy(&vOut, &vOutF, 16);
	printf("vf2h: %08x,%08x,%08x,%08x\n", vOut.x, vOut.y, vOut.z, vOut.w);
	vf2h(&vOutF, &vIn4); memcpy(&vOut, &vOutF, 16);
	printf("vf2h: %08x,%08x,%08x,%08x\n", vOut.x, vOut.y, vOut.z, vOut.w);

	vh2f(&vOutF, &vIn1); memcpy(&vOut, &vOutF, 16);
	printf("vh2f: %08x,%08x,%08x,%08x\n", vOut.x, vOut.y, vOut.z, vOut.w);
	vh2f(&vOutF, &vIn2); memcpy(&vOut, &vOutF, 16);
	printf("vh2f: %08x,%08x,%08x,%08x\n", vOut.x, vOut.y, vOut.z, vOut.w);
	vh2f(&vOutF, &vIn3); memcpy(&vOut, &vOutF, 16);
	printf("vh2f: %08x,%08x,%08x,%08x\n", vOut.x, vOut.y, vOut.z, vOut.w);
	vh2f(&vOutF, &vIn4); memcpy(&vOut, &vOutF, 16);
	printf("vh2f: %08x,%08x,%08x,%08x\n", vOut.x, vOut.y, vOut.z, vOut.w);

	vf2h(&vOutF, &vIn1); vh2f(&vOutF, &vOutF);
	printf("vf2h vh2f: %f,%f,%f,%f\n", vOutF.x, vOutF.y, vOutF.z, vOutF.w);
	vf2h(&vOutF, &vIn2); vh2f(&vOutF, &vOutF);
	printf("vf2h vh2f: %f,%f,%f,%f\n", vOutF.x, vOutF.y, vOutF.z, vOutF.w);
	vf2h(&vOutF, &vIn3); vh2f(&vOutF, &vOutF);
	printf("vf2h vh2f: %f,%f,%f,%f\n", vOutF.x, vOutF.y, vOutF.z, vOutF.w);
	vf2h(&vOutF, &vIn4); vh2f(&vOutF, &vOutF);
	printf("vf2h vh2f: %f,%f,%f,%f\n", vOutF.x, vOutF.y, vOutF.z, vOutF.w);
}


void checkVF2I() {
	static __attribute__ ((aligned (16))) ScePspFVector4 vIn[8] =
	{
		{0.0f, 0.1f, 0.5f, 0.9f}, 
		{1.0f, 1.1f, 1.5f, 1.9f},
		{2.0f, 2.5f, 3.5f, 1000.0f},
		{INFINITY, NAN, 0.0f, 0.0f},
		{-0.0f, -0.1f, -0.5f, -0.9f},
		{-1.0f, -1.1f, -1.5f, -1.9f},
		{-2.0f, -2.5f, -3.5f, -1000.0f},
		{-INFINITY, -NAN},
	};
	
	static __attribute__ ((aligned (16))) ScePspFVector4 vOutF =
	{0.0f, 0.0f, 0.0f, 0.0f};

	struct {int x,y,z,w;} vOut;
	int i;
	for (i = 0; i < 8; i++) {
		vf2id(&vOutF, &vIn[i]); memcpy(&vOut, &vOutF, 16);
		printf("vf2id: %i,%i,%i,%i\n", vOut.x, vOut.y, vOut.z, vOut.w);
		vf2in(&vOutF, &vIn[i]); memcpy(&vOut, &vOutF, 16);
		printf("vf2in: %i,%i,%i,%i\n", vOut.x, vOut.y, vOut.z, vOut.w);
		vf2iz(&vOutF, &vIn[i]); memcpy(&vOut, &vOutF, 16);
		printf("vf2iz: %i,%i,%i,%i\n", vOut.x, vOut.y, vOut.z, vOut.w);
		vf2iu(&vOutF, &vIn[i]); memcpy(&vOut, &vOutF, 16);
		printf("vf2iu: %i,%i,%i,%i\n", vOut.x, vOut.y, vOut.z, vOut.w);
	}
}

void checkVI2F() {
	static __attribute__ ((aligned (16))) int vIn1[4] =
	{0, 0xFFFFFFFF, 3, 0x80000000};
	static __attribute__ ((aligned (16))) int vIn2[4] =
	{-1, -2, -3, 0x10000};

	static __attribute__ ((aligned (16))) ScePspFVector4 vOut =
	{0.0f, 0.0f, 0.0f, 0.0f};

	vi2f(&vOut, vIn1);
	printf("vi2f: %f,%f,%f,%f\n", vOut.x, vOut.y, vOut.z, vOut.w);
	vi2f(&vOut, vIn2);
	printf("vi2f: %f,%f,%f,%f\n", vOut.x, vOut.y, vOut.z, vOut.w);
}

void vfpu_vc2i_s(int* in, int *out);
void vfpu_vuc2i_s(int* in, int *out);
void vfpu_vs2i_s(int* in, int *out);
void vfpu_vs2i_p(int* in, int *out);
void vfpu_vi2f_q(int* in, float *out);

void test_vc2i() {
	int in_i[4];
	int out_i[4];
	resetAllMatrices();
	
	in_i[0] = 0x01234567;
	in_i[1] = 0x89ABCDEF;
	in_i[2] = 0x01234567;
	in_i[3] = 0x89ABCDEF;

	memset(out_i, 0, sizeof(out_i)); vfpu_vc2i_s(in_i, out_i); printf("vfpu_vc2i_s: %08X, %08X, %08X, %08X\n", out_i[0], out_i[1], out_i[2], out_i[3]);
}

void test_vuc2i() {
	int in_i[4];
	int out_i[4];
	resetAllMatrices();
	
	in_i[0] = 0x01234567;
	in_i[1] = 0x89ABCDEF;
	in_i[2] = 0x01234567;
	in_i[3] = 0x89ABCDEF;

	memset(out_i, 0, sizeof(out_i)); vfpu_vuc2i_s(in_i, out_i); printf("vfpu_vuc2i_s: %08X, %08X, %08X, %08X\n", out_i[0], out_i[1], out_i[2], out_i[3]);

	in_i[0] = 0xFEDCBA89;
	in_i[1] = 0x00000000;
	in_i[2] = 0x00000000;
	in_i[3] = 0x00000000;

	memset(out_i, 0, sizeof(out_i)); vfpu_vuc2i_s(in_i, out_i); printf("vfpu_vuc2i_s: %08X, %08X, %08X, %08X\n", out_i[0], out_i[1], out_i[2], out_i[3]);
}

void test_vs2i() {
	int in_i[4];
	int out_i[4];
	resetAllMatrices();
	
	in_i[0] = 0x01234567;
	in_i[1] = 0x89ABCDEF;

	memset(out_i, 0, sizeof(out_i)); vfpu_vs2i_s(in_i, out_i); printf("vfpu_vs2i_s: %08X, %08X, %08X, %08X\n", out_i[0], out_i[1], out_i[2], out_i[3]);
	memset(out_i, 0, sizeof(out_i)); vfpu_vs2i_p(in_i, out_i); printf("vfpu_vs2i_p: %08X, %08X, %08X, %08X\n", out_i[0], out_i[1], out_i[2], out_i[3]);

  in_i[0] = 0xFCCF00FF;
	in_i[1] = 0x1100FF11;

	memset(out_i, 0, sizeof(out_i)); vfpu_vs2i_s(in_i, out_i); printf("vfpu_vs2i_s: %08X, %08X, %08X, %08X\n", out_i[0], out_i[1], out_i[2], out_i[3]);
	memset(out_i, 0, sizeof(out_i)); vfpu_vs2i_p(in_i, out_i); printf("vfpu_vs2i_p: %08X, %08X, %08X, %08X\n", out_i[0], out_i[1], out_i[2], out_i[3]);
}

void test_vi2f() {
	int in_i[4];
	float out_f[4];
	resetAllMatrices();
	
	in_i[0] = 1;
	in_i[1] = 10;
	in_i[2] = 9999;
	in_i[3] = 1693234;

	out_f[0] = out_f[1] = out_f[2] = out_f[3] = 0.0f; vfpu_vi2f_q(in_i, out_f); printf("vfpu_vi2f_q: %f, %f, %f, %f\n", out_f[0], out_f[1], out_f[2], out_f[3]);
}

int main(int argc, char *argv[]) {
	printf("Started\n");
	{
		test_vc2i();
		test_vuc2i();
		test_vs2i();
		test_vi2f();

		printf("checkVF2I:\n"); checkVF2I();
		printf("checkVI2F:\n"); checkVI2F();
		printf("checkHalf:\n"); checkHalf();
	}
	printf("Ended\n");

	return 0;
}
