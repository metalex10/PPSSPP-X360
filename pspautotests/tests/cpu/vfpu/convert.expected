Started
vfpu_vc2i_s: 67000000, 45000000, 23000000, 01000000
vfpu_vuc2i_s: 33B3B3B3, 22A2A2A2, 11919191, 00808080
vfpu_vuc2i_s: 44C4C4C4, 5D5D5D5D, 6E6E6E6E, 7F7F7F7F
vfpu_vs2i_s: 45670000, 01230000, 00000000, 00000000
vfpu_vs2i_p: 45670000, 01230000, CDEF0000, 89AB0000
vfpu_vs2i_s: 00FF0000, FCCF0000, CDEF0000, 89AB0000
vfpu_vs2i_p: 00FF0000, FCCF0000, FF110000, 11000000
vfpu_vi2f_q: 0.062500, 0.625000, 624.937500, 105827.125000
checkVF2I:
vf2id: 0,0,0,0
vf2in: 0,0,0,1
vf2iz: 0,0,0,0
vf2iu: 0,1,1,1
vf2id: 1,1,1,1
vf2in: 1,1,2,2
vf2iz: 1,1,1,1
vf2iu: 1,2,2,2
vf2id: 2,2,3,1000
vf2in: 2,2,4,1000
vf2iz: 2,2,3,1000
vf2iu: 2,3,4,1000
vf2id: 2147483647,2147483647,0,0
vf2in: 2147483647,2147483647,0,0
vf2iz: 2147483647,2147483647,0,0
vf2iu: 2147483647,2147483647,0,0
vf2id: 0,-1,-1,-1
vf2in: 0,0,0,-1
vf2iz: 0,0,0,0
vf2iu: 0,0,0,0
vf2id: -1,-2,-2,-2
vf2in: -1,-1,-2,-2
vf2iz: -1,-1,-1,-1
vf2iu: -1,-1,-1,-1
vf2id: -2,-3,-4,-1000
vf2in: -2,-2,-4,-1000
vf2iz: -2,-2,-3,-1000
vf2iu: -2,-2,-3,-1000
vf2id: -2147483648,2147483647,0,0
vf2in: -2147483648,2147483647,0,0
vf2iz: -2147483648,2147483647,0,0
vf2iu: -2147483648,2147483647,0,0
checkVI2F:
vi2f: 0.000000,-1.000000,3.000000,-2147483648.000000
vi2f: -1.000000,-2.000000,-3.000000,65536.000000
checkHalf:
vf2h: 3d333b33,63d14166,c0400000,47800000
vf2h: bd33bb33,e3d1c166,c0400000,47800000
vf2h: 41003e00,c480c300,c0400000,47800000
vf2h: 7c004300,7ffffc00,c0400000,47800000
vh2f: 44ccc000,3fecc000,44ccc000,3ff4c000
vh2f: 44ccc000,bfecc000,44ccc000,bff4c000
vh2f: 00000000,3ff80000,00000000,40040000
vh2f: 00000000,400c0000,00000000,7f800380
vf2h vh2f: 0.899902,1.299805,2.699219,1000.500000
vf2h vh2f: -0.899902,-1.299805,-2.699219,-1000.500000
vf2h vh2f: 1.500000,2.500000,-3.500000,-4.500000
vf2h vh2f: 3.500000,inf,-inf,nan
Ended
